cBournhonesque/lightyear
# Add delta compression

Fixes https://github.com/cBournhonesque/lightyear/issues/127

This PR adds delta-compression, which means that instead of serializing the entire component we only serialize a diff between two component states. It's only enabled for components currently.
This means:
- we can do partial component updates (i.e. if 10 fields are the same and one is updated, we only send that one field through the network)
- the packets should be more compressible since they contain values that are close to 0
- some types might leverage delta-compression well such as InputTypes ?
- without delta compression, we serialize the component once and send this to all clients. With delta-compression, we have to store in memory the past value of the component, so that we can compute diffs from it. The diffs might be different for each client, so we have to serialize once for each client. Basically we use more CPU and memory but less bandwidth.

In this PR, we also start refactoring the protocol to add more type-erasure. We didn't add the delta-compression updates with a typed API because that would require specialization (we need one system for `<C: Component>` and one for `<C: Component + Diffable>`); so instead we fully use type-erasure for all delta-compression related functions. This is a good first step towards (maybe) having a single system that iterates through all archetypes instead of one system per component.
(also the `ComponentIndex` might be useful in the future to efficiently find archetypes that contain a given component that we want to replicate!)

TODO:

- [x] call Drop on the stored `NonNull<u8>` when we drop them

- [x] add tick Clamping for the ticks stored in the delta-compression ack tracker (so that if it's been a long time, we just lose the value)
- [x] test that it works and add unit tests
- check bandwidth gains
- [x] make sure we can enable/disable delta-compression at runtime

- [x] maybe store the `NonNull<u8>` in a shared resource? we don't need to copy the data once per connection

In ConnectionManager:

- delta_compression_data: `HashMap<ComponentKind, HashMap<ReplicationGroupId, BtreeMap<Tick, Vec(Entity, NotNull)>>>`
- or should we try to store it in a component on the entity, but still access it in a type-erased manner?
  - i.e. when we add delta-compression, we register another component `PastValues` or `ConfirmedHistory`? Maybe use PastValues so that the component is untyped, and the values are `BtreeMap<Tick, NotNull>`
  - then on any entity that has a delta-compression component, we start by inserting the ConfirmedHistory component.
  - then during delta compression we fetch the component as a PtrMut, cast it to `PastValues`, add a value. Maybe with an AtomicBool counter that tracks how many clients might need this value?
(one problem is that the tick is per replication-group, so we would store it duplicated on each entity?)

In ReplicationSend, we need to:
- delete old ticks in the `PastValues` component when all clients have either ACK-ed or LOST (NACK) the message.
- compute diffs based on the last acked-tick. So we could keep a map `HashMap<ComponentKind, HashMap<ReplicationGroup, Tick>>` that contains the last acked tick for this client. So we can know which data we can compute the diff from. (and if there's none, just from base)


- Think about what happens on the receiver side:
  - OPTION 1: use the delta's NetId.
    - sender generates `DeltaMessage<C::Delta>` (which is registered for serialization) and serializes it using its NetId.
    - on the receiver, we have a `write_fn` associated with the Delta which deserializes the message, and apply the diffs to the main type (so the delta needs to have an associated type to the main type?).
    - same thing, the remove fn removes the main type?
    - or just have a mapping in the registry from the delta's net-id to the main net-id.
    - PRO: this looks like we dynamically use delta-compression or not with the same component. We can use the NetId to identify if we are using delta-compression or not.
  - OPTION 2: use the main component's NetId.
    - the delta-message is serialized with the main component's net-id.


- [x] xpbd examples are broken, the ball goes into the cube!
This seems to be caused by the line
`// self.group_channels.entry(group_id).or_default().send_tick = Some(bevy_tick);`
(set send_tick on send).
WHY?
